design-proposals/klip-32-sql-testing-tool.md

```diff
@@ -0,0 +1,284 @@
+INSERT INTO foo (rowtime, id, col1) VALUES (1, '1', 1);
+
+ASSERT DATA bar OFFSET 1 (rowtime, id, col1) VALUES (1, '1', 1);
+
+CREATE OR REPLACE STREAM bar AS SELECT * FROM foo WHERE col1 = 123;
+
+INSERT INTO foo (rowtime, id, col1) VALUES (2, '2', 2);
+INSERT INTO foo (rowtime, id, col1) VALUES (3, '3', 123);
+
+ASSERT DATA bar OFFSET 2 (rowtime, id, col1) VALUES (3, '3', 123);
+ASSERT STREAM bar (id VARCHAR KEY, col1 INT) WITH (kafka_topic='BAR', value_format='JSON');
+
+---------------------------------------------------------------------------------------------------
+--@test: dml - stream - change column
+--@expected.error: io.confluent.ksql.util.KsqlException
+--@expected.message: "Cannot REPLACE data source: DataSource '`BAR`' has schema ..."
+---------------------------------------------------------------------------------------------------
+
+CREATE STREAM foo (id VARCHAR KEY, col1 VARCHAR, col2 VARCHAR) 
+    WITH (kafka_topic='foo', value_format='JSON');
+
+-- the below operation is invalid because "col2" will be missing from the
+-- schema after renaming it to col3
+CREATE STREAM bar AS SELECT id, col1, col2 FROM foo;
+CREATE OR REPLACE STREAM bar AS SELECT id, col1, col2 AS col3 FROM foo;
+```
+
+### Test Structure
+
+YATT will accept as a parameter a directory containing testing files, and will run all the
+tests in the directory in a single JVM.
+
+Each test file can contain one or more tests, separated by the `--@test` directive. This will
+improve on a pain point of the existing ksql-test-runner, which requires three files for a single
+test. The test name is a concatenation of the test file time and the contents of the `--@test`
+directive.
+
+Then, the test can contain any number of interleaved statements that are supported by the testing
+tool. Data will be inserted into topics using `INSERT INTO` and asserts will be executed using the
+new `ASSERT` syntax (described below).
+
+There will be additional "meta" directives that will allow for testing functionality like expected
+exceptions.
+
+### The `ASSERT` Statement
+
+`ASSERT` is the primary way to ensure conditions in YATT, and will be parsed and implemented 
+using ksqlDB's `AstBuilder` to make sure YATT can leverage the existing expression parsing support. 
+
+At first, `ASSERT` statements will fail if they are sent to a production server, but we may consider
+allowing `ASSERT` in the future in order to allow REPL-based testing/experimentation.
+
+#### `ASSERT DATA`
+
+The `ASSERT DATA` statement asserts that data exists at an optionally provided offset (or otherwise 
+sequential based on the last `ASSERT DATA`) with the given values. If the `ASSERT DATA` does not
+specify certain columns they will be ignored (allowing users to assert only subset of the columns
+match what they expect).
+
+The values will be created in the same way that `INSERT INTO` creates values:
+```sql
+ASSERT DATA source_name [OFFSET at_offset] ( { column_name } [, ...] ) VALUES ( value [, ...] );
```

Comment:
What about asserting data in internal topics (e.g., repartition topics) as some QTT tests currently do?

Reply:
I discussed this with @big-andy-coates and we came up with two conclusions:

1. this shouldn't be necessary anymore because the historical tests capture all internal topic data and assert that it's the same, so you shouldn't need to manually assert that it stays the same
1. if you _really_ wanted to, you could declare it as a stream/table and use the `ASSERT DATA` calls